File: cpp/src/selection_sort.cpp
#include "algorithms.h"
#include <thread>
#include <chrono>
#include <iostream>

// Selection Sort with visualization
// Time Complexity: O(n^2) - quadratic
// Space Complexity: O(1) - constant
//
// How it works:
// 1. Find the smallest element in the unsorted part
// 2. Swap it with the first unsorted element
// 3. Move the boundary between sorted and unsorted
void selectionSort(std::vector<int>& array, Visualizer& viz) {
    int n = array.size();
    std::vector<bool> sorted(n, false);
    int totalComparisons = 0;
    int totalSwaps = 0;
    auto startTime = std::chrono::high_resolution_clock::now();

    std::cout << "\n========================================\n";
    std::cout << "Starting Selection Sort\n";
    std::cout << "========================================\n";
    std::cout << "Array size: " << n << "\n";
    std::cout << "Worst case: O(n^2) = " << (n * n) << " comparisons\n";
    std::cout << "Best case: O(n^2) = " << (n * n) << " comparisons\n";
    std::cout << "========================================\n\n";

    for (int i = 0; i < n - 1; i++) {
        int minIndex = i;
        int passComparisons = 0;

        std::cout << "Pass " << (i + 1) << "/" << (n - 1) << " - Finding smallest in unsorted part... ";

        // Find the minimum element in unsorted part
        for (int j = i + 1; j < n; j++) {
            totalComparisons++;
            passComparisons++;

            // Check for quit
            if (viz.shouldQuit()) return;

            // Visualize comparison
            viz.draw(array, minIndex, j, sorted);
            viz.playTone(array[j]);

            if (array[j] < array[minIndex]) {
                minIndex = j;
            }

            // Delay so we can see the visualization
            std::this_thread::sleep_for(std::chrono::milliseconds(viz.getDelayMs()));
        }

        // Swap the found minimum element with the first element
        if (minIndex != i) {
            std::swap(array[i], array[minIndex]);
            totalSwaps++;
            std::cout << passComparisons << " comparisons, 1 swap\n";
        } else {
            std::cout << passComparisons << " comparisons, 0 swaps (already in place)\n";
        }

        // Mark this position as sorted
        sorted[i] = true;

        // Show the swap
        viz.draw(array, i, minIndex, sorted);
        std::this_thread::sleep_for(std::chrono::milliseconds(viz.getDelayMs() * 3));
    }

    // Mark last element as sorted
    sorted[n - 1] = true;

    auto endTime = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);

    std::cout << "\n========================================\n";
    std::cout << "Selection Sort Complete!\n";
    std::cout << "========================================\n";
    std::cout << "Total comparisons: " << totalComparisons << "\n";
    std::cout << "Total swaps: " << totalSwaps << "\n";
    std::cout << "Time elapsed: " << duration.count() << "ms\n";
    std::cout << "Time complexity: O(n^2)\n";
    std::cout << "Space complexity: O(1)\n";
    std::cout << "========================================\n";

    // Final visualization showing all bars in green
    viz.draw(array, -1, -1, sorted);
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
}

File: rust/src/algorithms/selection_sort.rs
use macroquad::prelude::*;
use std::thread;
use std::time::{Duration, Instant};

use crate::algorithms::SortAlgorithm;
use crate::visualizer::{AudioSystem, draw_array};

// Selection Sort with visualization
// Time Complexity: O(n^2) - quadratic
// Space Complexity: O(1) - constant
//
// How it works:
// 1. Find the smallest element in the unsorted part
// 2. Swap it with the first unsorted element
// 3. Move the boundary between sorted and unsorted
pub async fn selection_sort(array: &mut [usize], algorithm: SortAlgorithm, audio: &AudioSystem, array_size: usize, delay_ms: u64) {
    let n = array.len();
    let mut sorted = vec![false; n];
    let mut total_comparisons = 0;
    let mut total_swaps = 0;
    let start_time = Instant::now();

    println!("\n========================================");
    println!("Starting Selection Sort");
    println!("========================================");
    println!("Array size: {}", n);
    println!("Worst case: O(n^2) = {} comparisons", n * n);
    println!("Best case: O(n^2) = {} comparisons", n * n);
    println!("========================================\n");

    for i in 0..n - 1 {
        let mut min_index = i;
        let mut pass_comparisons = 0;

        print!("Pass {}/{} - Finding smallest in unsorted part... ", i + 1, n - 1);

        // Find the minimum element in unsorted part
        for j in i + 1..n {
            total_comparisons += 1;
            pass_comparisons += 1;

            // Visualize comparison
            draw_array(array, Some(min_index), Some(j), &sorted, algorithm, array_size, delay_ms);
            next_frame().await;

            // Play tone for current element being checked
            audio.play_tone(array[j], array_size);

            if array[j] < array[min_index] {
                min_index = j;
            }

            // Delay so we can see the visualization
            thread::sleep(Duration::from_millis(delay_ms));
        }

        // Swap the found minimum element with the first element
        if min_index != i {
            array.swap(i, min_index);
            total_swaps += 1;
            println!("{} comparisons, 1 swap", pass_comparisons);
        } else {
            println!("{} comparisons, 0 swaps (already in place)", pass_comparisons);
        }

        // Mark this position as sorted
        sorted[i] = true;

        // Show the swap
        draw_array(array, Some(i), Some(min_index), &sorted, algorithm, array_size, delay_ms);
        next_frame().await;
        thread::sleep(Duration::from_millis(delay_ms * 3));
    }

    // Mark last element as sorted
    sorted[n - 1] = true;

    let duration = start_time.elapsed();

    println!("\n========================================");
    println!("Selection Sort Complete!");
    println!("========================================");
    println!("Total comparisons: {}", total_comparisons);
    println!("Total swaps: {}", total_swaps);
    println!("Time elapsed: {}ms", duration.as_millis());
    println!("Time complexity: O(n^2)");
    println!("Space complexity: O(1)");
    println!("========================================");

    // Final visualization showing all bars in green
    draw_array(array, None, None, &sorted, algorithm, array_size, delay_ms);
    next_frame().await;
    thread::sleep(Duration::from_millis(1000));
}
